lib/cfg.js

```diff
@@ -0,0 +1,383 @@
+    ++first;
+
+    var rangeBb = {nameId:functionName.substr(0,50) , start:first , end:null, actionPos:[]};
+    var result = [];
+
+    var newRangeWith = function (oldRange, nameId, start) {
+        return {nameId: nameId, start: start, actionPos: [], end: oldRange.end};
+    };
+
+    while(first != last) {
+        var inst = asmArr[first].text;
+        if(isEnd(inst)) {
+            rangeBb.end = first;
+            result.push(_.clone(rangeBb));
+            ++first;
+            //inst is expected to be .L*: where * in 1,2,...
+            rangeBb = newRangeWith(rangeBb, inst, first);
+        }
+        else if(isJmp(inst)) {
+                rangeBb.actionPos.push(first);
+        }
+        ++first;
+    }
+
+    rangeBb.end = last;
+    result.push(_.clone(rangeBb));
+    return result;
+};
+
+ControlFlowGraph.prototype.splitToCanonicalBasicBlock = function (basicBlock) {
+    var actionPos = basicBlock.actionPos;
+    var actPosSz = actionPos.length;
+    if (actionPos[actPosSz-1] + 1 == basicBlock.end){
+        --actPosSz;
+    }
+
+    if(actPosSz === 0)
+        return [{
+                    nameId: basicBlock.nameId,
+                    start: basicBlock.start,
+                    end: basicBlock.end
+               }];
+    else if(actPosSz == 1)
+            return [
+                {nameId: basicBlock.nameId,start: basicBlock.start,end:actionPos[0]+1},
+                {nameId: basicBlock.nameId+"@"+ (actionPos[0]+1),start: actionPos[0]+1,end: basicBlock.end}
+               ];
+    else {
+        var first = 0;
+        var last = actPosSz;
+        var blockName = basicBlock.nameId;
+        var tmp = {nameId:blockName, start:basicBlock.start, end:actionPos[first]+1 };
+        var result = [];
+        result.push(_.clone(tmp));
+        while(first != last-1) {
+            tmp.nameId = blockName + "@" + (actionPos[first]+1);
+            tmp.start = actionPos[first]+1;
+            ++first;
+            tmp.end = actionPos[first]+1;
+            result.push(_.clone(tmp));
+        }
+        
+        tmp = {nameId:blockName + "@" + (actionPos[first] + 1), start:actionPos[first] + 1, end:basicBlock.end };
+        result.push(_.clone(tmp));
+        
+        return result;
+
+    }
+
+};
+
+
+ControlFlowGraph.prototype.concatInstructions = function (asmArr, first, last) {
+  return _.chain(asmArr.slice(first, last))
+    .map(function (x) { return x.text.substr(0, 50); })
+    .value()
+    .join("\n");
+}; 
+
+
+ControlFlowGraph.prototype.makeNodes = function (asmArr, arrOfCanonicalBasicBlock) {
+    var node = {};
+    var nodes = [];
+    
+    _.each(arrOfCanonicalBasicBlock, _.bind(function(x){
+        logger.debug("node name:");
+        logger.debug(x.nameId);
+        node.id = x.nameId;
+        node.label = x.nameId + ((x.nameId.indexOf(":") !== -1)? "":":") +"\n"+
+                         this.concatInstructions(asmArr,x.start, x.end);
+        node.color = "#99ccff";//"#FFAFAF";
```

Comment:
Can you remove the commented out colour unless it's needed for something?

Reply:
done